components/raftstore/src/store/async_io/write.rs

```diff
@@ -0,0 +1,636 @@
+            message_metrics: Default::default(),
+            perf_context,
+        }
+    }
+
+    fn run(&mut self) {
+        let mut stopped = false;
+        while !stopped {
+            let mut handle_begin = None;
+
+            let mut first_time = true;
+            while self.batch.get_raft_size() < self.raft_write_size_limit {
+                let msg = if first_time {
+                    first_time = false;
+                    match self.receiver.recv() {
+                        Ok(msg) => {
+                            handle_begin = Some(Instant::now());
+                            msg
+                        }
+                        Err(_) => {
+                            stopped = true;
+                            break;
+                        }
+                    }
+                } else {
+                    match self.receiver.try_recv() {
+                        Ok(msg) => msg,
+                        Err(TryRecvError::Empty) => break,
+                        Err(TryRecvError::Disconnected) => {
+                            stopped = true;
+                            break;
+                        }
+                    }
+                };
+                match msg {
+                    WriteMsg::Shutdown => {
+                        stopped = true;
+                        break;
+                    }
+                    WriteMsg::WriteTask(task) => {
+                        self.metrics
+                            .task_wait
+                            .observe(duration_to_sec(task.send_time.saturating_elapsed()));
+                        self.batch.add_write_task(task);
+                    }
+                }
+            }
+
+            if self.batch.is_empty() {
+                continue;
+            }
+
+            STORE_WRITE_HANDLE_MSG_DURATION_HISTOGRAM
+                .observe(duration_to_sec(handle_begin.unwrap().saturating_elapsed()));
+
+            STORE_WRITE_TRIGGER_SIZE_HISTOGRAM.observe(self.batch.get_raft_size() as f64);
+
+            self.write_to_db();
+
+            self.metrics.flush();
+
+            // update config
+            if let Some(incoming) = self.cfg_tracker.any_new() {
+                self.raft_write_size_limit = incoming.raft_write_size_limit.0 as usize;
+                self.metrics.waterfall_metrics = incoming.waterfall_metrics;
+            }
+
+            STORE_WRITE_LOOP_DURATION_HISTOGRAM
+                .observe(duration_to_sec(handle_begin.unwrap().saturating_elapsed()));
+        }
+    }
+
+    fn write_to_db(&mut self) {
+        self.batch.before_write_to_db(&self.metrics);
+
+        fail_point!("raft_before_save");
+
+        if !self.batch.kv_wb.is_empty() {
+            let now = Instant::now();
+            let mut write_opts = WriteOptions::new();
+            write_opts.set_sync(true);
+            self.batch.kv_wb.write_opt(&write_opts).unwrap_or_else(|e| {
```

Comment:
Should also observe perf context.

Reply:
In the original code, these perf metrics are mixed with raft db which is very strange. Maybe we need another `PerfContextKind` to fix it(BTW, destroy process writes both kv db and raft db so it should also be fixed).
I think it's not very urgent. We can fix it in future PR.

Reply:
Then should add TODO.